nomad/fsm.go

```diff
@@ -374,7 +374,7 @@ func (n *nomadFSM) applyAllocUpdate(buf []byte, index uint64) interface{} {
 	// prior to being inserted into MemDB.
 	if j := req.Job; j != nil {
 		for _, alloc := range req.Alloc {
-			if alloc.Job == nil {
+			if alloc.Job == nil && !alloc.TerminalStatus() {
```

Comment:
Can we add a test in FSM where job is actually nil if the allocation is not for the currently registered job with the same ID?